### PR TITLE
fix: calling order of optional parameter

### PR DESCRIPTION
**Pull request by 1e1**
**! tested only on PHP 5.3 !**

While a route catch parameters in a certain order AND the relative method declares its parameters in another order AND all optional paramters are not setted,
the parameters are not mapped in the correct order.

Route: `#^/startups(?:/(?<provider>[\w\-]+)(?:/(?<domain>[\w\-]+))?)?/(?:§(?<page>\d+))?$#`
Method: `($page = 0, $provider = null, $domain = null)`
Url: `/startups/aProvider/aDomain`
-> 
call: `f('aProvider', 'aDomain', --null--)`

My user case occured in a controller which is extended by 2 psecific controllers.

### CODE REVIEW

**Comment by Hywan** [comment]
Still opened?

**Comment by Hywan** [comment]
@osaris: ping?

**Comment by osaris** [comment]
@1e1 ping ?

**Comment by 1e1** [comment]
Hi, 

I have not re-read the code. 
I remember that if I defined a method with optional parameters `C::m($year=null, $month=null, $day=null)`, then if I defined 2 routes: 
- `$router->get('date_fr', '/fr/(?<day>\d{2})?-(?<month>\d{2})?-(?<year>\d{4})?', 'C', 'm')`;
- `$router->get('date_en', '/en/(?<month>\d{2})?-(?<day>\d{2})?-(?<year>\d{4})?', 'C', 'm')`;
- `$router->get('date_iso', '/(?<year>\d{4})?-(?<month>\d{2})?-(?<day>\d{2})?', 'C', 'm')`;

Only `/2015-07-` will be correct.
`/fr/-07-2015` and `/fr/07--2015` will call `C::m(07, 2015, null)`

> "why parameter order has this side effect"
Even if the parameter is optional we have to set it with the default value, otherwise it is not present and the parameter list is sliced.

**Comment by osaris** [comment]
I have reproduced the problem, if the last part of the regex is empty then the ```$arguments``` array doesn't include values for missing parameters. For example, if you test ```/fr/--2015```  it works well because ```$arguments``` contains a value for each variables (empty for missing values). But if you test ```/fr/01--``` the ```$arguments``` array contains only a value for ```day```.

As invokeArgs use params by order not by name, then it affiliates the first and only params (day) to the first parameter of the method. 

Your patch is ok for me, it adds missing parameters to the ```$arguments```array so all parameters (even optionals) are passed to the called method.

**Comment by Hywan** [comment]
Is it possible to rebase the patch to `master` please? (`git rebase master` might do the job)

**Comment by 1e1** [comment]
Sorry for the delay. 
I have just updated the PR. 
Then we should run PHP-CS-fixer, there is a lot of Checkstyle mistakes.

**Comment by Hywan** [comment]
@1e1 Can I suggest you:
```php
$ composer install
$ vendor/bin/hoa devtools:cs --diff .
```
As mentionned here: http://hoa-project.net/En/Literature/Contributor/Guide.html#Preparing_a_patch.

Waiting for your go.

**Comment by 1e1** [comment]
No more changes. 
Go!

**Comment by Hywan** [comment]
@1e1 Please, can you rebase your commits now :-)? One commit will be fine.

**Comment by Hywan** [comment]
The error should probably exists on `Hoa\Dispatcher\ClassMethod`.
To merge this PR, I would like to know if this is possible to apply the same patch on `Hoa\Dispatcher\ClassMethod` but **also** to write a test for that particular case.

Is it possible?

File: Basic.php [hunk 1]
```diff
@@ -60,7 +60,7 @@ class Basic extends Dispatcher
      * @throws  \Hoa\Dispatcher\Exception
      */
     protected function resolve(
-        Array $rule,
+        array $rule,
```
Comment (Hywan) [inline Basic.php]: Keep `Array` here please.

**Comment by osaris** [comment]
@Hywan you asked for a review of this PR on IRC last week, It seems that :
* we lack a test to cover the problem addressed by this feature
* we lack a test to ensure the problem doesn't exists on ClassMethod dispatcher 
* we may lack a fix in the ClassMethod dispatcher if the test reveal that the bug also exists here

@1e1 can you please add a test for your PR and the same test for the ClassMethod disptacher ? We are here to help if you need !

**Comment by Hywan** [comment]
ping?

**Comment by Hywan** [comment]
I don't know how to apologize for the delay…

**Pull request merged**